src/material-examples/autocomplete-optgroup/autocomplete-optgroup-example.ts

```diff
@@ -0,0 +1,110 @@
+import {Component, OnInit} from '@angular/core';
+import {FormGroup, FormBuilder} from '@angular/forms';
+import {Observable} from 'rxjs';
+import {startWith, map} from 'rxjs/operators';
+/**
+ * @title Option groups autocomplete
+ */
+@Component({
+  templateUrl: './autocomplete-optgroup-example.html',
+  styleUrls: ['./autocomplete-optgroup-example.component.css'],
+})
+
+export class AutocompleteOptionGroupExample implements OnInit {
+  stateForm: FormGroup = this.fb.group({
+    stateGroup: '',
+  });
+
+  stateGroups: StateGroup[] = [{
+    letter: 'A',
+    names: ['Alabama', 'Alaska', 'Arizona', 'Arkansas']
+  }, {
+    letter: 'C',
+    names: ['California', 'Colorado', 'Connecticut']
+  }, {
+    letter: 'D',
+    names: ['Delaware']
+  }, {
+    letter: 'F',
+    names: ['Florida']
+  }, {
+    letter: 'G',
+    names: ['Georgia']
+  }, {
+    letter: 'H',
+    names: ['Hawaii']
+  }, {
+    letter: 'I',
+    names: ['Idaho', 'Illinois', 'Indiana', 'Iowa']
+  }, {
+    letter: 'K',
+    names: ['Kansas', 'Kentucky']
+  }, {
+    letter: 'L',
+    names: ['Louisiana']
+  }, {
+    letter: 'M',
+    names: ['Maine', 'Maryland', 'Massachusetts', 'Michigan',
+      'Minnesota', 'Mississippi', 'Missouri', 'Montana']
+  }, {
+    letter: 'N',
+    names: ['Nebraska', 'Nevada', 'New Hampshire', 'New Jersey',
+      'New Mexico', 'New York', 'North Carolina', 'North Dakota']
+  }, {
+    letter: 'O',
+    names: ['Ohio', 'Oklahoma', 'Oregon']
+  }, {
+    letter: 'P',
+    names: ['Pennsylvania']
+  }, {
+    letter: 'R',
+    names: ['Rhode Island']
+  }, {
+    letter: 'S',
+    names: ['South Carolina', 'South Dakota']
+  }, {
+    letter: 'T',
+    names: ['Tennessee', 'Texas']
+  }, {
+    letter: 'U',
+    names: ['Utah']
+  }, {
+    letter: 'V',
+    names: ['Vermont', 'Virginia']
+  }, {
+    letter: 'W',
+    names: ['Washington', 'West Virginia', 'Wisconsin', 'Wyoming']
+  }];
+
+  stateGroupOptions: Observable<StateGroup[]>;
+
+  constructor(private fb: FormBuilder) { }
+
+  ngOnInit() {
+    this.stateGroupOptions = this.stateForm.get('stateGroup').valueChanges
+      .pipe(
+        startWith(''),
+        map(val => this.filterGroup(val))
+      );
+  }
+
+  filterGroup(val: string): StateGroup[] {
+    if (val) {
+      return this.stateGroups
+        .map(group => ({ letter: group.letter, names: this._filter(group.names, val) }))
+        .filter(group => group.names.length > 0);
+    }
+
+    return this.stateGroups;
+  }
+
+  private _filter(opt: string[], val: string) {
+    const filterValue = val.toLowerCase();
+    return opt.filter(item => item.toLowerCase().startsWith(filterValue));
+  }
+}
+
+class StateGroup {
```

Comment:
You could replace this by an `Interface` and move this to the top of the file (after imports);